#!/usr/bin/env python

import subprocess
import sys

from script_commands import UserError, helps, run_subcommand


def tc(*arguments):
    """Run a tc command under sudo.

    :param arguments: The remainder of the command (leaving out tc).
    """
    subprocess.call(["sudo", "tc"] + arguments)


@helps(
    delay="Length of delay in milliseconds (each way).",
    port="Port to induce delay on.",
)
def start(delay=500, port=443):
    """Add artificial latency to the lo interface on the specified port."""
    qdisc_add = ["qdisc", "add", "dev", "lo"]
    tc(*qdisc_add, "root", "handle", "1:", "prio")
    tc(
        *qdisc_add,
        "parent",
        "1:3",
        "handle",
        "30:",
        "netem",
        "delay",
        "%dms" % delay,
    )
    filter_add_ip = ["filter", "add", "dev", "lo", "protocol", "ip"]
    tc(
        *filter_add_ip,
        "parent",
        "1:0",
        "prio",
        "3",
        "u32",
        "match",
        "ip",
        "dport",
        str(port),
        "0xffff",
        "flowid",
        "1:3",
    )
    tc(
        *filter_add_ip,
        "parent",
        "1:0",
        "prio",
        "3",
        "u32",
        "match",
        "ip",
        "sport",
        str(port),
        "0xffff",
        "flowid",
        "1:3",
    )


def stop():
    """Remove latency from the lo."""
    tc("qdisc", "del", "dev", "lo", "root")


subcommands = {
    "start": start,
    "stop": stop,
}


if __name__ == "__main__":
    try:
        run_subcommand(subcommands, sys.argv[1:])
    except UserError as e:
        sys.stderr.write(str(e) + "\n")
